File: 프로그래머스/3/12904. 가장 긴 팰린드롬/가장 긴 팰린드롬.cpp
#include <iostream>
#include <string>
using namespace std;
int dp[2500][2500];
string s;
int go(int left,int right){
    if(left>right){
        return 0;
    }
    if(left==right){
        return dp[left][right]=1;
    }
    if(dp[left][right]){
        return dp[left][right];
    }
    int l=go(left,right-1);
    int r=go(left+1,right);
    int mid=go(left+1,right-1);
    
    if(mid==right-left-1&&s[left]==s[right]){
        mid+=2;
    }
    return dp[left][right]=max(l,max(r,mid));
}
int solution(string s) {
    ::s=s;
    return go(0,s.length()-1);
}

File: 프로그래머스/3/12904. 가장 긴 팰린드롬/가장 긴 팰린드롬.java
class Solution{
    
    int[][] dp = new int[2500][2500];
    String s;
    
    public int solution(String s){
        this.s = s;
        return go(0, s.length() - 1);
    }
    
    int go(int n, int m) {
        if (n > m) {
            return 0;
        }
        if (n == m) {
            return 1;
        }
        if (dp[n][m] > 0) {
            return dp[n][m];
        }
        int left = go(n, m - 1);
        int right = go(n + 1, m);
        int mid = go(n + 1, m - 1);
        if (s.charAt(n) == s.charAt(m) && mid == m - n - 1) {
            mid += 2;
        }
        return dp[n][m] = Math.max(left, Math.max(right, mid));
    }
}